### ISSUE TITLE
Armeria Request and Response Metrics Questions

### ISSUE THREAD

**Issue by angela278** [line/armeria#5303]
Hi team, have a few questions regarding certain Armeria metrics.

1. What's the difference between armeria_request_duration and armeria_response_duration? Is there documentation somewhere describing each of the metrics? (apologies if it's on the website, we couldn't find it)
2. We're seeing high request durations despite our request timeout being set to 3 minutes. How is this possible?
![Screenshot 2023-11-14 at 10 18 25 PM](https://github.com/line/armeria/assets/5898426/5b491dce-9595-4517-8e45-04bb68122d40)

Thank you!

**Comment by trustin** [comment]
Hi! Thanks for the questions, @angela278.

> What's the difference between armeria_request_duration and armeria_response_duration? Is there documentation somewhere describing each of the metrics? 

Does this page answer your questions? https://armeria.dev/docs/advanced-structured-logging Please let us know if there's anything that's unclear.

> We're seeing high request durations despite our request timeout being set to 3 minutes. How is this possible?

According to the graph you attached, a request to `/v1/log` took 295 seconds which is way greater than 180 seconds (3 minutes), correct? This is actually very interesting because the request should be aborted and closed when the request is not consumed completely within `requestTimeoutMillis` (3 minutes). Is there any chance where ...

- your service or decorator reschedules the timeout using the methods such as `ServiceRequestContext.setRequestTimeout()`?
- you block any event loops, preventing Armeria from handling request timeouts on time?
- your machine is heavily loaded, preventing Armeria from handling request timeouts on time?

**Comment by angela278** [comment]
> Does this page answer your questions? https://armeria.dev/docs/advanced-structured-logging Please let us know if there's anything that's unclear.

Yes that page is great, thank you so much! One more clarification, what's the difference between armeria_server_requests and armeria_server_active_requests?

> According to the graph you attached, a request to /v1/log took 295 seconds which is way greater than 180 seconds (3 minutes), correct?

That is correct. I think out of the options you gave, _your machine is heavily loaded_ is the most likely one. At the time of the screenshot, we had pods with >2k connections and >15k requests (Our limits are set to 5000 `maxNumConnections` and 100 `maxNumRequestsPerConnection`). Could you go into more detail as to how a heavily loaded machine would prevent Armeria from handling request timeouts on time? We do see armeria_server_timeouts during this time as well with the cause as `RequestTimeoutException` (around 6.5k at the time in the screenshot).

**Comment by jrhee17** [comment]
> One more clarification, what's the difference between armeria_server_requests and armeria_server_active_requests?

`armeria.server.requests` is a counter which is incremented with every request, whereas `armeria.server.active.requests` is a gauge which records the in-flight "active" requests at a given time.

> Could you go into more detail as to how a heavily loaded machine would prevent Armeria from handling request timeouts on time?

This basically means the event loop is overloaded with tasks so that methods like `EventLoop#schedule` won't work as expected. If you suspect this, I believe there is a `NettyEventExecutorMetrics` in micrometer which exposes the number of pending tasks for event loops.

https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/netty4/NettyEventExecutorMetrics.java
